[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Stef0916/chemoinformatics-bioinformatics/blob/main/prediction-fluorination-strength/notebooks/train-test_split-3.ipynb)

In [2]:
!pip install openpyxl

In [1]:
import numpy as np
import pandas as pd

## Load Data

In [3]:
df1 = pd.read_excel('reduced_descriptors.xlsx')

In [4]:
df2 = pd.read_excel('F260_raw.xlsx')

In [5]:
df1

nSpiro    ATSC1c    ATSC3c    ATSC4c    ATSC5c    ATSC6c    ATSC8c  \
0         0 -0.165111 -0.059759  0.120854  0.066913 -0.013867 -0.050482   
1         0 -0.165111 -0.059759  0.120854  0.066913 -0.013867 -0.050482   
2         0 -0.252447 -0.005811  0.031723 -0.110729  0.237455  0.063804   
3         0 -0.252447 -0.005811  0.031723 -0.110729  0.237455  0.063804   
4         0 -0.164064 -0.069624  0.104539  0.053068  0.042946 -0.003314   
..      ...       ...       ...       ...       ...       ...       ...   
255       0 -0.356338  0.133605 -0.201203  0.087543  0.094537  0.001518   
256       0 -0.015593  0.000000  0.000000  0.000000  0.000000  0.000000   
257       0 -0.015593  0.000000  0.000000  0.000000  0.000000  0.000000   
258       0 -0.057496  0.000000  0.000000  0.000000  0.000000  0.000000   
259       0 -0.057496  0.000000  0.000000  0.000000  0.000000  0.000000   

       ATSC4dv     ATSC3d    ATSC4d  ...    AMID_N      JGI2      JGI3  \
0    50.150617  -6.640000 -2.511111  ...  0.104059  0.121864  0.073529   
1    50.150617  -6.640000 -2.511111  ...  0.104059  0.121864  0.073529   
2    28.227605 -10.269896 -1.041522  ...  0.094652  0.124183  0.067434   
3    28.227605 -10.269896 -1.041522  ...  0.094652  0.124183  0.067434   
4    77.398429  -7.358127  1.404040  ...  0.099120  0.127946  0.071181   
..         ...        ...       ...  ...       ...       ...       ...   
255  29.647959  -5.530612 -3.872449  ...  0.157364  0.086957  0.045000   
256   0.000000   0.000000  0.000000  ...  1.103553  0.000000  0.000000   
257   0.000000   0.000000  0.000000  ...  1.103553  0.000000  0.000000   
258   0.000000   0.000000  0.000000  ...  0.400000  0.000000  0.000000   
259   0.000000   0.000000  0.000000  ...  0.400000  0.000000  0.000000   

         JGI4      JGI5      JGI6      JGI7      JGI8      JGI9     JGI10  
0    0.041864  0.035791  0.018367  0.008433  0.004031  0.005625  0.000000  
1    0.041864  0.035791  0.018367  0.008433  0.004031  0.005625  0.000000  
2    0.044444  0.041419  0.020408  0.012427  0.010078  0.006944  0.004959  
3    0.044444  0.041419  0.020408  0.012427  0.010078  0.006944  0.004959  
4    0.045972  0.040638  0.017746  0.012804  0.007421  0.010417  0.008264  
..        ...       ...       ...       ...       ...       ...       ...  
255  0.049444  0.029684  0.018022  0.018353  0.012311  0.007049  0.010597  
256  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
257  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
258  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
259  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  

[260 rows x 88 columns]

In [6]:
df2

Name of the Reagent Reagent ID Number  \
0        N-fluoro-N-(phenylsulfonyl)benzenesulfonamide               A1a   
1        N-fluoro-N-(phenylsulfonyl)benzenesulfonamide               A1a   
2    N-fluoro-4-methoxy-N-(phenylsulfonyl)benzenesu...               A1b   
3    N-fluoro-4-methoxy-N-(phenylsulfonyl)benzenesu...               A1b   
4    N-fluoro-4-methyl-N-(phenylsulfonyl)benzenesul...               A1c   
..                                                 ...               ...   
255  (1R,2R,4S,5S)-2-((S)-acetoxy(6-methoxyquinolin...               F6c   
256                                Dinitrogen Fluoride                F7   
257                                Dinitrogen Fluoride                F7   
258                                  Ammonium Fluoride                F8   
259                                  Ammonium Fluoride                F8   

          Type of Reagent Solvent  \
0    N-Fluorosulfonimides    MeCN   
1    N-Fluorosulfonimides     DCM   
2    N-Fluorosulfonimides    MeCN   
3    N-Fluorosulfonimides     DCM   
4    N-Fluorosulfonimides    MeCN   
..                    ...     ...   
255     N-Fluoroammoniums     DCM   
256     N-Fluoroammoniums    MeCN   
257     N-Fluoroammoniums     DCM   
258     N-Fluoroammoniums    MeCN   
259     N-Fluoroammoniums     DCM   

                                                Smiles    FPD  
0    O=S(C1=CC=CC=C1)(N(S(C2=CC=C([H])C=C2)(=O)=O)F)=O  229.6  
1    O=S(C1=CC=CC=C1)(N(S(C2=CC=C([H])C=C2)(=O)=O)F)=O  242.4  
2     O=S(C1=CC=CC=C1)(N(S(C2=CC=C(OC)C=C2)(=O)=O)F)=O  231.2  
3     O=S(C1=CC=CC=C1)(N(S(C2=CC=C(OC)C=C2)(=O)=O)F)=O  244.2  
4      O=S(C1=CC=CC=C1)(N(S(C2=CC=C(C)C=C2)(=O)=O)F)=O  230.3  
..                                                 ...    ...  
255  F[N@+]1(C[C@H]2CC)[C@@H]([C@@H](OC(C)=O)C3=C(C...  250.7  
256                                            N#[N+]F  110.9  
257                                            N#[N+]F  112.3  
258                                       F[N+](F)(F)F  150.9  
259                                       F[N+](F)(F)F  153.0  

[260 rows x 6 columns]

In [7]:
data_to_split = pd.concat([df1, df2[' Type of Reagent'], df2['Solvent'], df2['Name of the Reagent'], df2['Smiles'], df2['FPD']], axis=1)

In [8]:
data_to_split

nSpiro    ATSC1c    ATSC3c    ATSC4c    ATSC5c    ATSC6c    ATSC8c  \
0         0 -0.165111 -0.059759  0.120854  0.066913 -0.013867 -0.050482   
1         0 -0.165111 -0.059759  0.120854  0.066913 -0.013867 -0.050482   
2         0 -0.252447 -0.005811  0.031723 -0.110729  0.237455  0.063804   
3         0 -0.252447 -0.005811  0.031723 -0.110729  0.237455  0.063804   
4         0 -0.164064 -0.069624  0.104539  0.053068  0.042946 -0.003314   
..      ...       ...       ...       ...       ...       ...       ...   
255       0 -0.356338  0.133605 -0.201203  0.087543  0.094537  0.001518   
256       0 -0.015593  0.000000  0.000000  0.000000  0.000000  0.000000   
257       0 -0.015593  0.000000  0.000000  0.000000  0.000000  0.000000   
258       0 -0.057496  0.000000  0.000000  0.000000  0.000000  0.000000   
259       0 -0.057496  0.000000  0.000000  0.000000  0.000000  0.000000   

       ATSC4dv     ATSC3d    ATSC4d  ...      JGI6      JGI7      JGI8  \
0    50.150617  -6.640000 -2.511111  ...  0.018367  0.008433  0.004031   
1    50.150617  -6.640000 -2.511111  ...  0.018367  0.008433  0.004031   
2    28.227605 -10.269896 -1.041522  ...  0.020408  0.012427  0.010078   
3    28.227605 -10.269896 -1.041522  ...  0.020408  0.012427  0.010078   
4    77.398429  -7.358127  1.404040  ...  0.017746  0.012804  0.007421   
..         ...        ...       ...  ...       ...       ...       ...   
255  29.647959  -5.530612 -3.872449  ...  0.018022  0.018353  0.012311   
256   0.000000   0.000000  0.000000  ...  0.000000  0.000000  0.000000   
257   0.000000   0.000000  0.000000  ...  0.000000  0.000000  0.000000   
258   0.000000   0.000000  0.000000  ...  0.000000  0.000000  0.000000   
259   0.000000   0.000000  0.000000  ...  0.000000  0.000000  0.000000   

         JGI9     JGI10       Type of Reagent  Solvent  \
0    0.005625  0.000000  N-Fluorosulfonimides     MeCN   
1    0.005625  0.000000  N-Fluorosulfonimides      DCM   
2    0.006944  0.004959  N-Fluorosulfonimides     MeCN   
3    0.006944  0.004959  N-Fluorosulfonimides      DCM   
4    0.010417  0.008264  N-Fluorosulfonimides     MeCN   
..        ...       ...                   ...      ...   
255  0.007049  0.010597     N-Fluoroammoniums      DCM   
256  0.000000  0.000000     N-Fluoroammoniums     MeCN   
257  0.000000  0.000000     N-Fluoroammoniums      DCM   
258  0.000000  0.000000     N-Fluoroammoniums     MeCN   
259  0.000000  0.000000     N-Fluoroammoniums      DCM   

                                   Name of the Reagent  \
0        N-fluoro-N-(phenylsulfonyl)benzenesulfonamide   
1        N-fluoro-N-(phenylsulfonyl)benzenesulfonamide   
2    N-fluoro-4-methoxy-N-(phenylsulfonyl)benzenesu...   
3    N-fluoro-4-methoxy-N-(phenylsulfonyl)benzenesu...   
4    N-fluoro-4-methyl-N-(phenylsulfonyl)benzenesul...   
..                                                 ...   
255  (1R,2R,4S,5S)-2-((S)-acetoxy(6-methoxyquinolin...   
256                                Dinitrogen Fluoride   
257                                Dinitrogen Fluoride   
258                                  Ammonium Fluoride   
259                                  Ammonium Fluoride   

                                                Smiles    FPD  
0    O=S(C1=CC=CC=C1)(N(S(C2=CC=C([H])C=C2)(=O)=O)F)=O  229.6  
1    O=S(C1=CC=CC=C1)(N(S(C2=CC=C([H])C=C2)(=O)=O)F)=O  242.4  
2     O=S(C1=CC=CC=C1)(N(S(C2=CC=C(OC)C=C2)(=O)=O)F)=O  231.2  
3     O=S(C1=CC=CC=C1)(N(S(C2=CC=C(OC)C=C2)(=O)=O)F)=O  244.2  
4      O=S(C1=CC=CC=C1)(N(S(C2=CC=C(C)C=C2)(=O)=O)F)=O  230.3  
..                                                 ...    ...  
255  F[N@+]1(C[C@H]2CC)[C@@H]([C@@H](OC(C)=O)C3=C(C...  250.7  
256                                            N#[N+]F  110.9  
257                                            N#[N+]F  112.3  
258                                       F[N+](F)(F)F  150.9  
259                                       F[N+](F)(F)F  153.0  

[260 rows x 93 columns]

In [9]:
data_to_split.to_excel('raw_data_to_split_into_train_test.xlsx', index=False)

## Stratified Split Based on "Type of Reagent"

On the publication, the author especifies that the splitting of the data into training and test set should ensure that:

1. The data from different **types of reagents** is propotional in both sets.
2. The combination **solvent-FPD** values for a particular molecule stays togethers either in the training or test set.

In [10]:
# Initialize the train ratio
train_ratio = 0.85

# Initialize empty DataFrames train and test data
train_data = pd.DataFrame(columns=data_to_split.columns)
test_data = pd.DataFrame(columns=data_to_split.columns)

# Group by ' Type of Reagent'
for reagent_type, reagent_type_group in data_to_split.groupby(' Type of Reagent'):

    # List of unique reagents in this group
    unique_reagents = reagent_type_group['Name of the Reagent'].unique()

    # Calculate the number of reagents to sample for training
    num_train_reagents = int(len(unique_reagents) * train_ratio)

    # Randomly sample reagents for the training set
    sampled_reagents = np.random.choice(unique_reagents, size=num_train_reagents, replace=False)

    # Add the corresponding data for these reagents to the training set
    train_data = pd.concat([train_data, reagent_type_group[reagent_type_group['Name of the Reagent'].isin(sampled_reagents)]])

    # Add the remaining data to the test set
    test_data = pd.concat([test_data, reagent_type_group[~reagent_type_group['Name of the Reagent'].isin(sampled_reagents)]])

# Reset index for clarity
train_data.reset_index(drop=True, inplace=True)
test_data.reset_index(drop=True, inplace=True)

train_data.shape, test_data.shape  # Display the shape of train and test data

((214, 93), (46, 93))

In [11]:
# Analyze the proportionality of "Type of Reagent"

# Proportions in the entire dataset
overall_proportions = data_to_split[' Type of Reagent'].value_counts(normalize=True)

# Proportions in the training set
train_proportions = train_data[' Type of Reagent'].value_counts(normalize=True)

# Proportions in the test set
test_proportions = test_data[' Type of Reagent'].value_counts(normalize=True)

# Combine the proportions into a single DataFrame
proportions_df = pd.DataFrame({
    'Overall': overall_proportions,
    'Train': train_proportions,
    'Test': test_proportions
})

proportions_df

Overall     Train      Test
N-Fluoropyridiniums   0.307692  0.308411  0.304348
N-Fluorosulfonimides  0.261538  0.261682  0.260870
N-Fluoroammoniums     0.161538  0.158879  0.173913
N-Fluorosulfonamides  0.153846  0.158879  0.130435
N-Fluorocarboxamides  0.076923  0.074766  0.086957
N-Fluoroheterocycles  0.038462  0.037383  0.043478

The proportions of the types of reagents is similar in the three datasets, the initial one containing the 260 molecules (130 x 2 solvents), the training and test set.

In [12]:
# Verify that "Name of the Reagent" values are not split between train and test sets

# Unique reagents in the training list
train_reagents = list(train_data['Name of the Reagent'].unique())

# Unique reagents in the test list
test_reagents = list(test_data['Name of the Reagent'].unique())

# Find reagents that might appear in both sets
common_reagents = [value for value in train_reagents if value in test_reagents]

common_reagents

[]

We verified that the same molecule in both solvents got sorted together to either the test or the training set.

In [13]:
train_data.to_excel('train_data.xlsx', index=False)

In [14]:
test_data.to_excel('test_data.xlsx', index=False)

In [15]:
train_data.head(6)

nSpiro    ATSC1c    ATSC3c    ATSC4c   ATSC5c    ATSC6c  ATSC8c    ATSC4dv  \
0      0 -0.096458  0.057844  0.031682  0.00859  0.006587     0.0  -4.584968   
1      0 -0.096458  0.057844  0.031682  0.00859  0.006587     0.0  -4.584968   
2      0 -0.038971 -0.007437  0.147682  0.01105 -0.018099     0.0 -17.693762   
3      0 -0.038971 -0.007437  0.147682  0.01105 -0.018099     0.0 -17.693762   
4      0 -0.059595  0.043280  0.076304  0.00934 -0.002160     0.0  -1.958400   
5      0 -0.059595  0.043280  0.076304  0.00934 -0.002160     0.0  -1.958400   

      ATSC3d    ATSC4d  ...  JGI6  JGI7  JGI8  JGI9  JGI10    Type of Reagent  \
0 -12.867200  3.145600  ...   0.0   0.0   0.0   0.0    0.0  N-Fluoroammoniums   
1 -12.867200  3.145600  ...   0.0   0.0   0.0   0.0    0.0  N-Fluoroammoniums   
2  -9.703214  2.109641  ...   0.0   0.0   0.0   0.0    0.0  N-Fluoroammoniums   
3  -9.703214  2.109641  ...   0.0   0.0   0.0   0.0    0.0  N-Fluoroammoniums   
4 -10.747200  1.785600  ...   0.0   0.0   0.0   0.0    0.0  N-Fluoroammoniums   
5 -10.747200  1.785600  ...   0.0   0.0   0.0   0.0    0.0  N-Fluoroammoniums   

   Solvent                                Name of the Reagent  \
0     MeCN  1-(chloromethyl)-4-fluoro-1,4-diazabicyclo[2.2...   
1      DCM  1-(chloromethyl)-4-fluoro-1,4-diazabicyclo[2.2...   
2     MeCN  1-fluoro-4-hydroxy-1,4-diazabicyclo[2.2.2]octa...   
3      DCM  1-fluoro-4-hydroxy-1,4-diazabicyclo[2.2.2]octa...   
4     MeCN  1-fluoro-4-methyl-1,4-diazabicyclo[2.2.2]octan...   
5      DCM  1-fluoro-4-methyl-1,4-diazabicyclo[2.2.2]octan...   

                       Smiles    FPD  
0  F[N+]1(CC2)CC[N+]2(CCl)CC1  229.8  
1  F[N+]1(CC2)CC[N+]2(CCl)CC1  225.5  
2    F[N+]1(CC2)CC[N+]2(O)CC1  226.8  
3    F[N+]1(CC2)CC[N+]2(O)CC1  222.3  
4    F[N+]1(CC2)CC[N+]2(C)CC1  232.1  
5    F[N+]1(CC2)CC[N+]2(C)CC1  228.1  

[6 rows x 93 columns]

In [16]:
test_data.head(6)

nSpiro    ATSC1c    ATSC3c    ATSC4c    ATSC5c    ATSC6c    ATSC8c  \
0      0 -0.027476  0.015512  0.049684  0.011963  0.000000  0.000000   
1      0 -0.027476  0.015512  0.049684  0.011963  0.000000  0.000000   
2      0 -0.212913  0.174961 -0.141682  0.030300  0.140330 -0.165645   
3      0 -0.212913  0.174961 -0.141682  0.030300  0.140330 -0.165645   
4      0 -0.356338  0.133605 -0.201203  0.087543  0.094537  0.001518   
5      0 -0.356338  0.133605 -0.201203  0.087543  0.094537  0.001518   

     ATSC4dv     ATSC3d    ATSC4d  ...      JGI6      JGI7      JGI8  \
0 -33.917355  -9.099174  2.082645  ...  0.000000  0.000000  0.000000   
1 -33.917355  -9.099174  2.082645  ...  0.000000  0.000000  0.000000   
2  -2.816327 -14.005831 -7.189088  ...  0.019097  0.014248  0.010354   
3  -2.816327 -14.005831 -7.189088  ...  0.019097  0.014248  0.010354   
4  29.647959  -5.530612 -3.872449  ...  0.018022  0.018353  0.012311   
5  29.647959  -5.530612 -3.872449  ...  0.018022  0.018353  0.012311   

       JGI9     JGI10    Type of Reagent  Solvent  \
0  0.000000  0.000000  N-Fluoroammoniums     MeCN   
1  0.000000  0.000000  N-Fluoroammoniums      DCM   
2  0.009062  0.010597  N-Fluoroammoniums     MeCN   
3  0.009062  0.010597  N-Fluoroammoniums      DCM   
4  0.007049  0.010597  N-Fluoroammoniums     MeCN   
5  0.007049  0.010597  N-Fluoroammoniums      DCM   

                                 Name of the Reagent  \
0  1,4-difluoro-1,4-diazabicyclo[2.2.2]octane-1,4...   
1  1,4-difluoro-1,4-diazabicyclo[2.2.2]octane-1,4...   
2  (1R,2R,4S,5S)-1-fluoro-2-((S)-hydroxy(6-methox...   
3  (1R,2R,4S,5S)-1-fluoro-2-((S)-hydroxy(6-methox...   
4  (1R,2R,4S,5S)-2-((S)-acetoxy(6-methoxyquinolin...   
5  (1R,2R,4S,5S)-2-((S)-acetoxy(6-methoxyquinolin...   

                                              Smiles    FPD  
0                           F[N+]1(CC2)CC[N+]2(F)CC1  223.8  
1                           F[N+]1(CC2)CC[N+]2(F)CC1  218.9  
2  O[C@@H](C1=C(C=C(OC)C=C2)C2=NC=C1)[C@@H]3[N@@+...  249.2  
3  O[C@@H](C1=C(C=C(OC)C=C2)C2=NC=C1)[C@@H]3[N@@+...  252.7  
4  F[N@+]1(C[C@H]2CC)[C@@H]([C@@H](OC(C)=O)C3=C(C...  248.0  
5  F[N@+]1(C[C@H]2CC)[C@@H]([C@@H](OC(C)=O)C3=C(C...  250.7  

[6 rows x 93 columns]